File: mplssim/display.py
"""Central presentation display-name registry.

Internal IDs (PE1, P5, L11, D2, scenario keys, …) are the contract used by
the pretrained model, configs, tests and result files — they are NEVER
renamed. This module maps them to presentation labels: Turkish city names on
a FICTIONAL scaled national backbone (not a real operator topology — the UI
carries that disclaimer).

Everything user-facing derives its labels from this one registry, served to
both frontends via GET /api/display.
"""

from __future__ import annotations

from typing import Any

CITY_NAMES: dict[str, str] = {
    "PE1": "İstanbul", "PE2": "İzmir", "PE3": "Bursa", "PE4": "Antalya",
    "P1": "Eskişehir", "P2": "Ankara", "P3": "Konya", "P4": "Bolu",
    "P5": "Kayseri", "P6": "Adana", "P7": "Gaziantep", "P8": "Samsun",
    "A1": "Sivas", "A2": "Malatya",
    "PE5": "Trabzon", "PE6": "Erzurum", "PE7": "Diyarbakır", "PE8": "Van",
}

SCENARIO_NAMES: dict[str, str] = {
    "full_day": "Normal National Traffic Day",
    "morning_ramp": "Morning Business Ramp-Up",
    "evening_peak": "Evening Streaming Peak",
    "night_consolidation": "Quiet Night Operations",
    "flash_crowd": "Major Live Event Traffic Surge",
    "link_failure": "Ankara–Kayseri Backbone Failure",
    "demand_forecast_error": "Unforecast Demand Surge",
    "deceptive_local_optimum": "Hidden Shared Bottleneck",
    "demo_evening": "Guided Operator Demonstration",
    "random_day": "Randomized Training Day",
    "ood_double_failure": "Two-Link Regional Outage",
    "overload_stress": "Nationwide Capacity Overload",
}

CLASS_NAMES: dict[str, str] = {
    "voice": "voice", "video": "video", "vpn": "enterprise VPN",
    "besteffort": "consumer internet", "bulk": "bulk data", "critical": "critical services",
}

DISCLAIMER = ("Fictional scaled national backbone for demonstration — "
              "not a real operator topology.")

GLOSSARY: dict[str, str] = {
    "Traffic demand": "A group of data travelling between two cities.",
    "Route": "The sequence of cities carrying that traffic.",
    "Link utilization": "How much of a connection's capacity is being used.",
    "SLA problem": "A service is experiencing too much delay or packet loss.",
    "AI Advisor": "A trained policy that recommends which traffic route to change.",
    "Total reward": "The experiment's combined score for delivery, congestion, "
                    "service quality, and route stability. It is a simulation "
                    "score, not money or a direct industry KPI.",
    "Fast reroute": "Immediate use of a backup path after a link fails.",
    "Traditional controller": "A rule-based method that reacts to current congestion.",
}


def scale_mbps(value_mbps: float, factor: float) -> float:
    """Display-only scaling used by Presentation Mode's 'scaled national
    backbone' view. Applied identically to loads AND capacities so that
    utilization is invariant; never applied to delay, loss, SLA counts,
    reroutes or reward. The frontend mirrors this exact rule."""
    return value_mbps * factor


def city(router_id: str) -> str:
    return CITY_NAMES.get(router_id, router_id)


def path_label(routers: list[str] | tuple[str, ...]) -> str:
    """'İstanbul → Eskişehir → Kayseri → Samsun → Erzurum'"""
    return " → ".join(city(r) for r in routers)


def link_label(a: str, z: str) -> str:
    """'Ankara–Kayseri link' (technical id shown separately in tooltips)."""
    return f"{city(a)}–{city(z)} link"


def demand_label(src: str, dst: str, cls: str) -> str:
    """'İstanbul → Erzurum video traffic'"""
    return f"{city(src)} → {city(dst)} {CLASS_NAMES.get(cls, cls)} traffic"


def scenario_label(key: str) -> str:
    return SCENARIO_NAMES.get(key, key)


def display_bundle(topology: Any = None) -> dict[str, Any]:
    """Everything the frontends need, in one payload (GET /api/display)."""
    links = {}
    if topology is not None:
        for ld in topology.link_defs.values():
            links[ld.id] = {
                "label": link_label(ld.a, ld.z),
                "technical": f"{ld.a}–{ld.z}, {ld.id}",
            }
    return {
        "cities": CITY_NAMES,
        "scenarios": SCENARIO_NAMES,
        "classes": CLASS_NAMES,
        "links": links,
        "disclaimer": DISCLAIMER,
        "glossary": GLOSSARY,
    }

File: frontend/js/display.js
// Presentation label registry, client side.
//
// The server owns the mapping (mplssim/display.py, served at GET /api/display).
// This module caches it once at boot and turns internal IDs into the labels the
// audience sees. Internal IDs (PE1, P5, L11, D2, scenario keys) are NEVER
// renamed — they are the contract with the pretrained model, the configs, the
// tests and the committed result files. City names are display only.

const EMPTY = { cities: {}, scenarios: {}, classes: {}, links: {},
                disclaimer: "", glossary: {} };

let reg = EMPTY;

export async function loadDisplay() {
  const r = await fetch("/api/display");
  if (!r.ok) throw new Error(`/api/display returned ${r.status}`);
  reg = { ...EMPTY, ...(await r.json()) };
  return reg;
}

export const registry = () => reg;
export const disclaimer = () => reg.disclaimer;
export const glossary = () => reg.glossary;

/** "PE1" -> "İstanbul" (falls back to the raw ID if unmapped). */
export const city = (routerId) => reg.cities[routerId] || routerId;

/** ["PE1","P1","P5"] -> "İstanbul → Eskişehir → Kayseri". */
export const pathLabel = (routers) =>
  (routers || []).map(city).join(" → ");

/** Short chain for tight spaces: "İstanbul → … → Erzurum". */
export function pathShort(routers, maxHops = 4) {
  const p = routers || [];
  if (p.length <= maxHops) return pathLabel(p);
  return `${city(p[0])} → … → ${city(p[p.length - 1])}`;
}

/**
 * The transit cities only — "Eskişehir → Kayseri → Sivas". Two candidate
 * routes for the same demand share their endpoints, so the transit chain is
 * what actually distinguishes them in one line of the decision tape.
 */
export function pathVia(routers) {
  const mid = (routers || []).slice(1, -1);
  return mid.length ? mid.map(city).join(" → ") : "a direct link";
}

/** "L11" -> "Ankara–Kayseri link". */
export const linkLabel = (linkId) =>
  (reg.links[linkId] && reg.links[linkId].label) || linkId;

/** "L11" -> "P2–P5, L11" — the technical detail line, never the headline. */
export const linkTechnical = (linkId) =>
  (reg.links[linkId] && reg.links[linkId].technical) || linkId;

/** "Ankara–Kayseri link (P2–P5, L11)" for engineer-facing dropdowns. */
export const linkFull = (linkId) => `${linkLabel(linkId)} (${linkTechnical(linkId)})`;

export const className = (cls) => reg.classes[cls] || cls;

/** "İstanbul → Erzurum video traffic" — the plain-language demand name. */
export const demandLabel = (src, dst, cls) =>
  `${city(src)} → ${city(dst)} ${className(cls)} traffic`;

/** Same, with the internal ID appended for engineer-facing dropdowns. */
export const demandFull = (d) =>
  `${demandLabel(d.src, d.dst, d.class ?? d.cls)} (${d.id})`;

/** Scenario key -> "Guided Operator Demonstration". */
export const scenarioLabel = (key) => reg.scenarios[key] || key;

/** Human name for a controller, used in scoreboards and story text. */
const ALGO_LABELS = {
  rl: "AI Advisor", static: "Fixed routing", greedy: "Traditional controller",
  cspf: "CSPF re-optimizer", random: "Random baseline",
};
export const algoLabel = (a) => ALGO_LABELS[a] || a;

/** Technical name, for the engineering console. */
const ALGO_TECH = {
  rl: "RL (MaskablePPO)", static: "static shortest path", greedy: "greedy (util-aware)",
  cspf: "CSPF periodic reopt", random: "random floor",
};
export const algoTech = (a) => ALGO_TECH[a] || a;

/**
 * Resolve the undirected link ID joining two adjacent routers, using the
 * technical field ("P2–P5, L11") the server already sends. Used to name a
 * congested hop in story text without a second topology lookup.
 */
export function linkBetween(a, z) {
  for (const [id, info] of Object.entries(reg.links)) {
    const t = String(info.technical || "");
    if (t.startsWith(`${a}–${z},`) || t.startsWith(`${z}–${a},`)) return id;
  }
  return null;
}
